SO-4202: fix merge performance issue when merging huge amount of changes into a branch

cmark opened the pull request.

cmark committed:
SO-4202: load commits in batches of 20 to reduce memory constraint...

...when comparing changes between to branches with large result sets

cmark committed:
SO-4202: replace String.format with String.join...

...in ObjectId and RevisionCompareDetail to improve performance of
compare and merge processes.

cmark committed:
SO-4202: unrelated String.format replacements, cleanups

cmark committed:
Merge remote-tracking branch 'origin/7.x' into issue/SO-4202-merge_performance_issue

cmark committed:
SO-4202 #resolve

Fix merge performance issue when merging huge amount of changes from a
branch into another (speed up conflict detection).

apeteri commented:
`RevisionRangePathQueryTest` failed in the GitHub build for this PR:

```
readRange(com.b2international.index.revision.RevisionRangePathQueryTest)  Time elapsed: 0.128 s  <<< ERROR!
java.lang.ArrayIndexOutOfBoundsException: Index 0 out of bounds for length 0
	at com.b2international.index.revision.DefaultRevisionIndex.read(DefaultRevisionIndex.java:101)
	at com.b2international.index.revision.BaseRevisionIndexTest.search(BaseRevisionIndexTest.java:161)
	at com.b2international.index.revision.RevisionRangePathQueryTest.readRange(RevisionRangePathQueryTest.java:57)
```

apeteri commented on commons/com.b2international.index/src/com/b2international/index/revision/RevisionCompareDetail.java:
```diff
@@ -27,6 +27,8 @@
  */
 public final class RevisionCompareDetail {
 
+	private static final String PROPERTY_CHANGE_KEY_TEMPLATE = "_";
```
This is now a separator, not a String template. Please rename the constant

apeteri commented on commons/com.b2international.index/src/com/b2international/index/revision/RevisionCompareDetail.java:
```diff
@@ -40,6 +42,9 @@
 	private final String fromValue;
 	private final String value;
 	
+	@JsonIgnore
```
The property also has a &commat;`JsonIgnore` annotation.

apeteri commented on commons/com.b2international.index/src/com/b2international/index/revision/StagingArea.java:
```diff
@@ -673,43 +673,41 @@ void merge(RevisionBranchRef fromRef, RevisionBranchRef toRef, boolean squash, R
-								conflicts.add(new ChangedInSourceAndTargetConflict(sourceChange.getValue().getObject(), sourceChangeDiff.convert(conflictProcessor), targetChangeDiff.convert(conflictProcessor)));
-							} else {
+					if (sourcePropertyChanges != null) {
+						for (Entry<String, RevisionCompareDetail> sourceChange : Iterables.consumingIterable(sourcePropertyChanges.entrySet())) {
```
It no longer seems to be necessary to consume Map entries here 🤔 We're not doing anything with any leftover entries in `sourcePropertyChanges` (or indirectly via `sourcePropertyChangesByObject`).

apeteri commented on commons/com.b2international.index/src/com/b2international/index/revision/StagingArea.java:
```diff
@@ -673,43 +673,41 @@ void merge(RevisionBranchRef fromRef, RevisionBranchRef toRef, boolean squash, R
+					if (sourcePropertyChanges != null) {
+						for (Entry<String, RevisionCompareDetail> sourceChange : Iterables.consumingIterable(sourcePropertyChanges.entrySet())) {
+							final RevisionPropertyDiff sourceChangeDiff = new RevisionPropertyDiff(sourceChange.getValue().getProperty(), sourceChange.getValue().getFromValue(), sourceChange.getValue().getValue());
+							final RevisionCompareDetail targetPropertyChange = targetPropertyChanges.remove(sourceChange.getKey());
```
Same here.

apeteri reviewed (approved):
👍

cmark committed:
SO-4202: make property conflict detection algorithm more readable

Remove source,target prop change Maps from byObject Map before checking
for conflicts to slowly release allocated memory.

cmark committed:
SO-4202: rename constant to SEPARATOR

cmark committed:
SO-4202: use `String.join()` in RevisionIndex.toRevisionRange
